Move to ioredis for streaming

gmemstr opened the pull request:
As discovered in https://github.com/mastodon/mastodon/pull/26571, the current streaming server uses a Redis client that is incompatible with some nice-to-haves when it comes to Redis, specifically Sentinel and Cluster. This PR moves to [ioredis][1], which does support these features and is generally better supported.

I have done my best to test these changes, and following the [migration guide][2] it _seems_ fairly straightforward, but there may be some edge cases I haven't caught.

[1]: https://github.com/redis/ioredis
[2]: https://github.com/redis/ioredis/wiki/Migrating-from-node_redis

gmemstr committed:
Move to ioredis for streaming

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -33,21 +33,14 @@ const redisUrlToClient = async (defaultConfig, redisUrl) => {
-        path: redisUrl.slice(7),
-      },
-    }));
+    client = new Redis(redisUrl.slice(7));
```
Does this change mean that the other redis settings that come via config/env variables can't be overridden?

gmemstr replied:
I actually might refactor this a bit - are these specific ones we need to come through? I think ioredis supports everything we'd need https://github.com/redis/ioredis/blob/v4/API.md#new-redisport-host-options We'd just need to tweak the assignment where needed.

ThisIsMissEm replied:
Honestly, I'm not sure! I think everyone I know uses network sockets for connecting to redis, instead of unix sockets.

Main thing to look out for is being able to pass the password separately from the connection string, and setting any connection timeouts, I think

gmemstr replied:
Slightly tweaked this function and moved most of the configuration to the `redisConfigFromEnv()` function - had to shuffle some things around to work with the new library's options. Should be all good now :)

ThisIsMissEm reviewed (approved):
I honestly expected more changes for some reason, but this looks good. I've one question, but aside from that, nice work!

@ClearlyClaire this would be a good candidate for 4.2 if possible, as it fixes the issue with streaming halting when there's a network disruption between streaming & redis.

gmemstr committed:
Rework streaming redis configuration options

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -148,14 +132,19 @@ const pgConfigFromEnv = (env) => {
+    host: env.REDIS_HOST || '127.0.0.1',
+    port: env.REDIS_PORT || 6379,
+    options: {
+	  path: path,
+	  db: env.REDIS_DB || 0,
+      password: env.REDIS_PASSWORD || undefined,
+	}
```
Indentation here is a bit wonky, be sure to check this with prettier/linting (`npm run lint` from the root, or `prettier --check streaming/index.js`)

```suggestion
  	  path: path,
  	  db: env.REDIS_DB || 0,
      password: env.REDIS_PASSWORD || undefined,
	  }
```

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -148,14 +132,19 @@ const pgConfigFromEnv = (env) => {
+    host: env.REDIS_HOST || '127.0.0.1',
+    port: env.REDIS_PORT || 6379,
+    options: {
+	  path: path,
```
what happens if `path` is null? does it get ignored? Or should we only set `redisParams.path if path`?

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -148,14 +132,19 @@ const pgConfigFromEnv = (env) => {
-    },
-    database: env.REDIS_DB || 0,
-    password: env.REDIS_PASSWORD || undefined,
+    host: env.REDIS_HOST || '127.0.0.1',
+    port: env.REDIS_PORT || 6379,
```
ditto, if path, then I don't think host & port should be set?

gmemstr replied:
ioredis will use `path` over `host` and `port` itself :)

ThisIsMissEm replied:
Let's document that in the code, just to avoid confusion later. e.g., have host, port and path lines next to each other, and then describe that behaviour.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -29,25 +29,9 @@ log.level = process.env.LOG_LEVEL || 'verbose';
  */
 const redisUrlToClient = async (defaultConfig, redisUrl) => {
```
we could also change this method to just be `redisClient = async (environment) => {` and do `const { redisParams, redisUrl, redisPrefix } = redisConfigFromEnv(environment);` inside this function.

Then the two lines at ~170/190, become:

```diff
-  const redisSubscribeClient = await redisUrlToClient(redisParams, redisUrl);
-  const redisClient = await redisUrlToClient(redisParams, redisUrl);
+  const redisSubscribeClient = await redisClient(process.env);
+  const redisClient = await redisClient(process.env);
```

ThisIsMissEm reviewed (commented):
Looking better! Still questions around handling of the config.

gmemstr committed:
Clean up Redis client function, remove manual redis prefix

gmemstr commented:
Pushed up some changes - appreciate the review! Also factored out the `redisPrefix` in favour of `ioredis`'s own `keyPrefix` option. Hopefully that's okay.

gmemstr committed:
Re-add redisUrl

ThisIsMissEm commented:
> Pushed up some changes - appreciate the review! Also factored out the `redisPrefix` in favour of `ioredis`'s own `keyPrefix` option. Hopefully that's okay.

Nice, but if you're only substituting one variable you don't need the template syntax, e.g., `(channelId, ...)` instead of
``` 
(`${channelId}`, ...)
```

gmemstr committed:
Remove uneccesary string interpolation

gmemstr commented:
@ThisIsMissEm Good point - fixed :)

gmemstr committed:
Correct issue with ioredis callbacks

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -24,30 +24,12 @@ dotenv.config({
-    }));
-  }
-
+const newRedisClient = async (environment) => {
```
Maybe `createRedisClient`?

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -150,23 +132,24 @@ const redisConfigFromEnv = (env) => {
 
-  if (redisNamespace) {
-    redisParams.namespace = redisNamespace;
+  if (env.REDIS_URL && env.REDIS_URL.startsWith('unix://')) {
```
This does still need a line of documentation "Setting redisParams.path overrides using host & port"

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -280,7 +261,7 @@ const startServer = async () => {
    * @param {string} channel
    */
-  const onRedisMessage = (message, channel) => {
+  const onRedisMessage = (channel, message) => {
```
Make sure you update the documentation comment above, I believe the lines need to be ordered correctly

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -312,7 +293,8 @@ const startServer = async () => {
       log.verbose(`Subscribe ${channel}`);
-      redisSubscribeClient.subscribe(channel, onRedisMessage);
+      redisSubscribeClient.subscribe(channel)
+	  redisSubscribeClient.on("message", onRedisMessage);
```
Does each subscribe require a "on message"? Or is this applying at client level? If so, you can lift this up to when you initialise the redis client for subscriptions, otherwise you'll be getting many invocations of the same callback each message.

ThisIsMissEm replied:
I think this should just be done once.

gmemstr replied:
I think you're right, it per client.

ThisIsMissEm reviewed (changes requested):
The onRedisMessage only needs to be registered once, as it's per client, not per topic.

gmemstr committed:
Cleanup of docstrings, hoist event assignment of subscription client

gmemstr commented on streaming/index.js:
```diff
@@ -294,6 +276,7 @@ const startServer = async () => {
 
     callbacks.forEach(callback => callback(json));
   };
+  redisSubscribeClient.on("message", onRedisMessage);
```
Thank you again @ThisIsMissEm for the reviews. I brought this up and out of the actual subscribe loop, but I'm not sure if I want to bring this up higher - I'd have to move the entire assignment of `onRedisMessage` further up so I opted to instead place the `.on()` here.

gmemstr committed:
Merge branch 'main' into ioredis

gmemstr committed:
Correctly set redis key prefix and path

gmemstr commented:
Tracked it down, had the  path and keyPrefix set in the wrong place - now correctly sets under the `.options.` object.

gmemstr committed:
Flatten redisParams

gmemstr commented:
I've run the workflows here https://github.com/gmemstr/mastodon/actions/runs/5985350688 with `DEBUG=ioredis:*` to capture all the logs from ioredis. It seems everything is connecting fine, but no messages are actually coming through. Currently looking at it from a prefix perspective, but functionally these should be identical. So I'm a little lost at this point.

gmemstr committed:
Correctly add : to redis namespace

gmemstr commented on streaming/index.js:
```diff
@@ -147,7 +147,7 @@ const redisConfigFromEnv = (env) => {
 
   if (redisNamespace) {
-    redisParams.keyPrefix = redisNamespace;
+    redisParams.keyPrefix = `${redisNamespace}:`;
```
Of course as soon as I post that I realise what was missing.

gmemstr replied:
This was correct - but only partially. ioredis doesn't automatically prefix `subscribe()` and such.

gmemstr replied:
Still not quite there yet.

gmemstr committed:
Manually prefix subscribe and unsubscribe

ioredis prefixes set(), get() etc. automatically, but does not prefix
subscribe and unsubscribe. This may be something we ask them upstream.

gmemstr committed:
Entirely revert keyPrefix

keyPrefix doesn't seem to work in all cases, and seems to entirely
ignore the subscription side of things. I'd still like to track down
the root cause but for now I've reverted the use of it.

gmemstr commented:
After much fighting, I decided to just revert the keyPrefix stuff. It's easier. I've managed to get the tests passing locally so I'm calling it here.

ClearlyClaire commented on streaming/index.js:
```diff
@@ -617,7 +602,7 @@ const startServer = async () => {
 
     subscribe(`${redisPrefix}${accessTokenChannelId}`, listener);
-    subscribe(`${redisPrefix}${systemChannelId}`, listener);
+    subscribe('${redisPrefix}${systemChannelId}', listener);
```
Wait, why this change?

gmemstr replied:
Oops, typo. Fixed and pushed.

gmemstr committed:
Correct quotation mark typo

gmemstr committed:
Whitespace

ClearlyClaire commented on streaming/index.js:
```diff
@@ -24,30 +24,15 @@ dotenv.config({
+  const { redisParams, redisUrl } = config;
+  let client = new Redis(redisParams);
+  if (redisUrl) {
+	client = new Redis(redisUrl, redisParams);
```
Indentation issue as well. Not sure why this isn't caught by the CI tests :thinking:

ClearlyClaire commented on streaming/index.js:
```diff
@@ -312,7 +297,7 @@ const startServer = async () => {
     if (subs[channel].length === 0) {
       log.verbose(`Subscribe ${channel}`);
-      redisSubscribeClient.subscribe(channel, onRedisMessage);
+      redisSubscribeClient.subscribe(channel)
```
Missing semicolon, another style issue I'm not sure why it isn't caught.

gmemstr replied:
I'm also a little confused as to how these weren't caught by linting, I would've expected this to be picked up. Might be worth opening a ticket/PR reviewing the linting rules in place.

gmemstr committed:
Linting

ClearlyClaire reviewed (approved):
Looks good overall, but I'm not acquainted to `ioredis`. We also probably want a *lot* of careful testing on this.

gmemstr commented:
For what it's worth, alongside testing this on my local machine, I've also applied the patch to the production instance I help run and haven't encountered any issues with it thus far.

ThisIsMissEm reviewed (changes requested):
Just one comment that needs fixing is the `new Redis` one; ideally we should handle the subscribe/unsubscribe command errors, even though those should be rare (we can just use `log.error` lines for those errors for now).

Some additional notes below:

Something to be mindful of is that whilst redis is down, the node.js process will queue up commands via ioredis's offline queue; this could cause a difficult situation if redis goes down and clients haven't correctly implemented reconnection with jitter and backoff, where you manage to bring up redis, thousands of clients attempt to resubscribe at once, taking redis offline, memory backs up, and node.js crashes due to OOM, and then you've gotta restart your entire infrastructure (we saw this exact failure mode at Pusher circa November 2011). In theory, by using `enableOfflineQueue: false` we could mitigate this issue.

Right now, there isn't actually anything in the websocket protocol that allows for a response to a subscribe failing, there's just an `{ error: string }` for errors meaning they're pretty opaque and don't correctly structure a reason, so a client wouldn't be able to know if the subscription actually failed (and we'd have no way of pipelining that back to the client that requested the subscription), so in the future we'll likely want to rework that, such that a subscribe call on the API could return back why and which subscription failed to the client.

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -24,30 +24,15 @@ dotenv.config({
+  const { redisParams, redisUrl } = config;
+  let client = new Redis(redisParams);
+  if (redisUrl) {
+    client = new Redis(redisUrl, redisParams);
```
You can actually write this as setting `redisUrl` to null or undefined if it's not present, due to how ioredis parses the options (null and undefined values are ignored), so you don't actually need a conditional here.

```
// The following are functionally equivalent:

new Redis(null, { ... })
new Redis({ ... })
new Redis({ ... }, null)
```

Just note that the options are applied left to right, so if you did `new Redis(redisParams, redisUrl)` then the `redisUrl` would take precedence over the `redisParams` options.

gmemstr replied:
I had this originally, but I think I reverted it while working through some of the subscription issues. Will swap back and test

gmemstr committed:
Merge branch 'main' into ioredis

gmemstr committed:
Utilise ioredis ccallback for subscription, simplify client creation

gmemstr commented on streaming/index.js:
```diff
@@ -312,7 +297,7 @@ const startServer = async () => {
 
     if (subs[channel].length === 0) {
       log.verbose(`Subscribe ${channel}`);
-      redisSubscribeClient.subscribe(channel, onRedisMessage);
```
And as a bonus, we can use `count` to get an accurate number for our metrics! :)

ThisIsMissEm replied:
Let's not change the metrics right now..

gmemstr commented:
@ThisIsMissEm Cheers - I do agree that we could implement better handling of failed subscriptions, I'll open a ticket later today to bookmark that issue.

As for the offline queue - I'm tempted to add this change to this PR, and then maybe look at reverting it/tweak it when I add Redis Sentinel support in #26571

gmemstr commented:
Fair enough - happy to leave this PR as is.

ClearlyClaire commented:
Am I right in understanding that the offline queue is a feature `ioredis` has that the previous library did not have, and that if we disabled it, it would be no worse than the current situation, but if we keep it enabled, it would work better in case of a short disconnect, but cause issues if the disconnect got longer?

If so, I'd advocate for disabling the offline queue for now, and investigating how to improve upon that in the future.

gmemstr commented:
From what I can tell, the previous library did actually do this (or, was supposed to. I haven't double checked it actually worked, just going off the codebase https://github.com/redis/node-redis/blob/294cbf8367295ac81cbe51ce2932493ab80493f1/packages/client/lib/client/index.ts#L53-L57 / https://github.com/redis/node-redis/blob/master/docs/client-configuration.md?plain=1). So I think we're safe to leave it on.

vmstan commented:
I've been running this on vmst.io since yesterday afternoon and haven't noticed any issues, but I'm going to apply the changes made in the last 12 hours now.

ThisIsMissEm commented:
> Am I right in understanding that the offline queue is a feature `ioredis` has that the previous library did not have, and that if we disabled it, it would be no worse than the current situation, but if we keep it enabled, it would work better in case of a short disconnect, but cause issues if the disconnect got longer?
> 
> If so, I'd advocate for disabling the offline queue for now, and investigating how to improve upon that in the future.

@ClearlyClaire yes, correct.

gmemstr committed:
Handle unsubscribe error

gmemstr commented on streaming/index.js:
```diff
@@ -312,8 +294,14 @@ const startServer = async () => {
+          log.error(`Error subscribing to ${channel}`);
+        }
+        else {
+          redisSubscriptions.set(count);
```
Done!

ClearlyClaire commented:
> @ClearlyClaire yes, correct.

As @gmemstr pointed out, the client we currently use does seem to maintain an offline queue: https://github.com/redis/node-redis/blob/master/docs/FAQ.md#what-happens-when-the-network-goes-down

Does it significantly differs from ioredis'?

ThisIsMissEm commented:
> > @ClearlyClaire yes, correct.
> 
> As @gmemstr pointed out, the client we currently use does seem to maintain an offline queue: https://github.com/redis/node-redis/blob/master/docs/FAQ.md#what-happens-when-the-network-goes-down
> 
> Does it significantly differs from ioredis'?

Thinking a bit more about this, the failure mode we saw at pusher was a bit different to the failure mode we'd see on the streaming server. As the only commands we're sending to redis from streaming are `un/subscribe` and `set`, the volume of data flowing between streaming and redis is very asymmetrical: a significant factor more data flows from redis to streaming than streaming to redis.

In the failure scenario at pusher, we had a significant amount of data flowing in both directions, which caused all services to back up in memory.

So a failure in redis is more likely to take down mastodon-web or the sidekiq instances than it is to take down streaming (given that one interaction amplifies into at least 2 redis publishes, but on the extreme side up to N publishes).

So we'd be better looking to ensure that the publishes don't results in queued data on the ruby side; the streaming server should be fine.

We'll see further improvements by moving from the `set` based "should we publish to this topic" to using the bullt-in `pubsub numsub` command, which would mitigate the counters being off (causing unnecessary publishes)

So for now I think we're fine leaving the offline queue enabled in streaming, once we move to pubsub numsub, we'd be able to disable it, as it'd just be subscribe calls which would fail (and we'll need a failure mode in the streaming protocol for that)

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -334,8 +322,14 @@ const startServer = async () => {
+        else {
+          redisSubscriptions.set(count);
+        }
+	  });
```
```suggestion
  	  });
```

gmemstr committed:
Update streaming/index.js

Co-authored-by: Emelia Smith <ThisIsMissEm@users.noreply.github.com>

gmemstr committed:
Update index.js

ThisIsMissEm commented on streaming/index.js:
```diff
@@ -334,8 +322,14 @@ const startServer = async () => {
+          redisSubscriptions.set(count);
+        }
+	  });
       delete subs[channel];
```
It's arguable if this should be done in the callback too, as if we unsubscribe but it fails we've not really unsubscribed — I'm fine with it either way though

ThisIsMissEm reviewed (changes requested):
Looks good, just one whitespace change needed

ThisIsMissEm reviewed (approved):
I see you just pushed a fix for the whitespace

demize commented:
Sorry to bump a closed PR; I just upgraded to 4.2.0-rc2 and ran into a bit of a (documentation, really) issue caused by this change.

My config did not have a `REDIS_URL` variable in it, only `REDIS_HOST`, `REDIS_PORT`, and `REDIS_PASSWORD`. The way the connection is built now, however, requires a REDIS_URL variable to be provided, lest you get this error:

```
node:internal/errors:478
    ErrorCaptureStackTrace(err);
    ^
TypeError [ERR_INVALID_ARG_TYPE]: The "url" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:387:5)
    at validateString (node:internal/validators:162:11)
    at Url.parse (node:url:168:3)
    at Object.urlParse [as parse] (node:url:155:13)
    at createRedisClient (/home/mastodon/mastodon/streaming/index.js:41:22)
    at startServer (/home/mastodon/mastodon/streaming/index.js:213:38)
    at Object.<anonymous> (/home/mastodon/mastodon/streaming/index.js:1556:1)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)
    at Module.load (node:internal/modules/cjs/loader:1074:32) {
  code: 'ERR_INVALID_ARG_TYPE'
}
```

The `REDIS_URL` variable is not mentioned in the upgrade instructions or provided in .env.production.sample, so this took a bit of troubleshooting before I could restore my streaming service. Given failure to update your config to include `REDIS_URL` is a fatal error, it'd probably make sense to mention this in the upgrade instructions and add the variable to the sample config file?

demize commented:
(Disregard my deleted comment; might not be as relevant as I thought...)
